src/parallel-rustc.md

```diff
@@ -1,34 +1,110 @@
 is run to combine all the codegen units together into one binary. This process
 occurs in the `rustc_codegen_ssa::base` module.
 
+## Data Structures
+
+The underlying thread-safe data-structures used in the parallel compiler
+can be found in the `rustc_data_structures::sync` module. These data structures 
+are implemented diferently depending on whether `parallel-compiler` is true.
+
+| data structure                   | parallel                                            | non-parallel |
+| -------------------------------- | --------------------------------------------------- | ------------ |
+| Lrc                              | std::sync::Arc | std::rc::Rc |
+| Weak                             | std::sync::Weak                                     | std::rc::Weak |
+| Atomic{Bool}/{Usize}/{U32}/{U64} | std::sync::atomic::Atomic{Bool}/{Usize}/{U32}/{U64} | (std::cell::Cell<bool/usize/u32/u64>) |
+| OnceCell                         | std::sync::OnceLock                                 | std::cell::OnceCell |
+| Lock\<T> | (parking_lot::Mutex\<T>) | (std::cell::RefCell) |
+| RwLock\<T> | (parking_lot::RwLock\<T>) | (std::cell::RefCell) |
+| MTRef<'a, T> | &'a T | &'a mut T |
+| MTLock\<T> | (Lock\<T>) | (T) |
+| ReadGuard | parking_lot::RwLockReadGuard | std::cell::Ref |
+| MappedReadGuard | parking_lot::MappedRwLockReadGuard | std::cell::Ref |
+| WriteGuard | parking_lot::RwLockWriteGuard | std::cell::RefMut |
+| MappedWriteGuard | parking_lot::MappedRwLockWriteGuard | std::cell::RefMut |
+| LockGuard | parking_lot::MutexGuard | std::cell::RefMut |
+| MappedLockGuard | parking_lot::MappedMutexGuard | std::cell::RefMut |
+| MetadataRef | [`OwningRef<Box<dyn Erased + Send + Sync>, [u8]>`][OwningRef] | [`OwningRef<Box<dyn Erased>, [u8]>`][OwningRef] |
+
+- There are currently a lot of global data structures that need to be made
+  thread-safe. A key strategy here has been converting interior-mutable
+  data-structures (e.g. `Cell`) into their thread-safe siblings (e.g. `Mutex`).
+
+### WorkLocal
+
+`WorkLocal` is a special data structure implemented for parallel compiler.
+It holds worker-locals values for each thread in a thread pool. You can only 
+access the worker local value through the Deref impl on the thread pool it 
+was constructed on. It will panic otherwise. 
+
+`WorkLocal` is used to implement the `Arena` allocator in the parallel 
+environment, which is critical in parallel queries. Its implementation 
+is located in the `rustc-rayon-core::worker_local` module. However, in the 
+non-parallel compiler, it is implemented as `(OneThread<T>)`, whose `T` 
+can be accessed directly through `Deref::deref`.
+
+## Parallel Iterator
+
+The parallel iterators provided by the [`rayon`] crate are efficient 
+ways to achieve parallelization. The current nightly rustc uses (a custom 
+fork of) [`rayon`] to run tasks in parallel. The custom fork allows the 
+execution of DAGs of tasks, not just trees.
```

Comment:
I'm not sure this is the only/primary reason (just not familiar with the reasons for the fork). Off hand I don't immediately see why rayon wouldn't work for a DAG.

IIRC, the fork was also adjusting some of the worker start/stop logic to use jobserver tokens; my recollection is that was a relatively iffy impl that sometimes saw tokens get lost.

Reply:
The rationale for using rayon's fork here comes from the statement in the original version. I don't quite understand it myself, so I'll drop this statement here for now and mark it as `(more information wanted)`. I'll improve it in the future